### PR TITLE
tDOT Deliveries

### PR DESCRIPTION
**Pull request by shengda**
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/872

### CODE REVIEW

**Review by alxs — CHANGES_REQUESTED** [review]
Thanks for the delivery @shengda. Please see my comments below. For the article, as usual feel free to send it to grantsPR@web3.foundation with grants@web3.foundation in cc for proofreading and cross-promotion.

File: deliveries/tdot-milestone-3.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+>
+> If there is anything particular about any of the deliverables we or a future reader should know, use the respective `Notes` column.
+
+| Number | Deliverable | Link | Notes |
+| ------------- | ------------- | ------------- |------------- |
+| 0b. | Documentation |[architecture](https://github.com/nutsfinance/stable-asset/blob/09b447d33291fb1507862898684a3279c4320ed3/architecture.png)||
+| 0c.  | Testing |[xcm](https://github.com/nutsfinance/stable-asset/blob/09b447d33291fb1507862898684a3279c4320ed3/lib/stable-asset-xcm/src/tests.rs) [stable-asset](https://github.com/nutsfinance/stable-asset/blob/09b447d33291fb1507862898684a3279c4320ed3/lib/stable-asset/src/tests.rs)||
+| 0d.  | Docker |[Acala](https://github.com/AcalaNetwork/Acala/blob/ad240e9b96d4338a66fe7daad5bf53d8bb6a25f8/scripts/Dockerfile) [bifrost](https://github.com/nutsfinance/bifrost/blob/f0cba77760cf7e9b4576f6a255c6496edd36aad0/Dockerfile)||
+| 0e.  | Article |[Article]()||
```
[the remark this answers is not in the document]
Comment (shengda) [reply deliveries/tdot-milestone-3.md]: I have updated the PR with the article.

https://dingshengda.medium.com/explaining-tdot-6a8cfe999f47

Do I still need to send the article to the Emails provided?

File: deliveries/tdot-milestone-3.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+* **Application Document:** [tDOT](https://github.com/w3f/Grants-Program/blob/master/applications/tdot.md).
+* **Milestone Number:** 3
+
+> Please provide a list of all deliverables of the milestone extracted from the initial application and a link to the deliverable itself. Ideally all links inside the below table should include a commit hash, which will be used for testing. If you don't provide a commit hash, we will work off the default branch of your repository. Thus, if you plan on continuing work after delivery, we suggest you create a separate branch for either the delivery or your continuing work.
+>
+> If there is anything particular about any of the deliverables we or a future reader should know, use the respective `Notes` column.
+
+| Number | Deliverable | Link | Notes |
+| ------------- | ------------- | ------------- |------------- |
+| 0b. | Documentation |[architecture](https://github.com/nutsfinance/stable-asset/blob/09b447d33291fb1507862898684a3279c4320ed3/architecture.png)||
```
[the remark this answers is not in the document]
Comment (shengda) [reply deliveries/tdot-milestone-3.md]: Link and README are updated in PR.

File: deliveries/tdot-milestone-3.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (alxs) [reply deliveries/tdot-milestone-3.md]: Many thanks. No, this is optional.

File: deliveries/tdot-milestone-3.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+| 0c.  | Testing |[xcm](https://github.com/nutsfinance/stable-asset/blob/09b447d33291fb1507862898684a3279c4320ed3/lib/stable-asset-xcm/src/tests.rs) [stable-asset](https://github.com/nutsfinance/stable-asset/blob/09b447d33291fb1507862898684a3279c4320ed3/lib/stable-asset/src/tests.rs)||
```
[the remark this answers is not in the document]
Comment (ukby1234) [reply deliveries/tdot-milestone-3.md]: this is the coverage I got: 
```
running 13 tests
test mock::__construct_runtime_integrity_test::runtime_integrity_tests ... ok
test tests::create_pool_successful ... ok
test tests::redeem_proportion_pool_not_found ... ok
test tests::mint_over_limit ... ok
test tests::mint_pool_not_found ... ok
test tests::redeem_proportion_over_limit ... ok
test tests::update_limit_pool_not_found ... ok
test tests::redeem_single_pool_not_found ... ok
test tests::redeem_single_over_limit ... ok
test tests::update_limit_successful ... ok
test tests::mint_successful ... ok
test tests::redeem_single_successful ... ok
test tests::redeem_proportion_successful ... ok

test result: ok. 13 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.06s

Aug 31 21:51:27.771  INFO cargo_tarpaulin::report: Coverage Results:
|| Uncovered Lines:
|| src/lib.rs: 258, 266, 279-281, 288-293, 297, 308, 310-311, 318, 330, 339, 353, 371-372, 375-376
|| src/mock.rs: 137, 181, 185, 189, 193, 197, 203, 210-212, 230, 248
|| src/tests.rs: 82, 128, 188, 254
|| src/weights.rs: 52, 54-55, 57, 59-60, 62, 64-65, 67, 69-70, 72, 74-75
|| Tested/Total Lines:
|| src/lib.rs: 127/150 +0.00%
|| src/mock.rs: 49/61 +0.00%
|| src/tests.rs: 155/159 -0.05%
|| src/weights.rs: 0/15 +0.00%
|| 
85.97% coverage, 331/385 lines covered, -0.10844825277816028% change in coverage
```
Not sure where the 23.9% came from, but added more tests.

File: deliveries/tdot-milestone-3.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+| 0d.  | Docker |[Acala](https://github.com/AcalaNetwork/Acala/blob/ad240e9b96d4338a66fe7daad5bf53d8bb6a25f8/scripts/Dockerfile) [bifrost](https://github.com/nutsfinance/bifrost/blob/f0cba77760cf7e9b4576f6a255c6496edd36aad0/Dockerfile)||
```
[the remark this answers is not in the document]
Comment (ukby1234) [reply deliveries/tdot-milestone-3.md]: Added and updated the Dockerfile. The reason why I didn't include it before is that this chain doesn't utilize the XCM feature so it only provides a proof-of-concept. 
For the XCM part, we actually integrated with two chains: Acala and bifrost and set up a testnet.

[the remark this answers is not in the document]
Comment (alxs) [reply deliveries/tdot-milestone-3.md]: Thanks for the clarification. That makes sense and the updated Dockerfile worked.

File: deliveries/tdot-milestone-3.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (alxs) [reply deliveries/tdot-milestone-3.md]: Many thanks. Sorry about that, the wrong coverage report must have been due to a bug in an older version of `cargo-tarpaulin` - I updated it and now get similar numbers.

File: deliveries/tdot-milestone-1.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+* **Milestone Number:** 1
+
+> Please provide a list of all deliverables of the milestone extracted from the initial application and a link to the deliverable itself. Ideally all links inside the below table should include a commit hash, which will be used for testing. If you don't provide a commit hash, we will work off the default branch of your repository. Thus, if you plan on continuing work after delivery, we suggest you create a separate branch for either the delivery or your continuing work.
+>
+> If there is anything particular about any of the deliverables we or a future reader should know, use the respective `Notes` column.
+
+| Number | Deliverable | Link | Notes |
+| ------------- | ------------- | ------------- |------------- |
+| 0b. | Documentation |[architecture](https://github.com/nutsfinance/stable-asset/blob/affe16c004003307bfa9c5e3db733d78e1c9f33e/architecture.png)||
+| 0c.  | Testing |[code](https://github.com/nutsfinance/stable-asset/blob/affe16c004003307bfa9c5e3db733d78e1c9f33e/lib/stable-asset-xcm/src/tests.rs)||
```
Comment (alxs) [inline deliveries/tdot-milestone-1.md:19]: Could you please link specifically to the testing described in this milestone "comprehensive tests that covers stable asset pool management functionalities"?

File: deliveries/tdot-milestone-2.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+* **Milestone Number:** 2
+
+> Please provide a list of all deliverables of the milestone extracted from the initial application and a link to the deliverable itself. Ideally all links inside the below table should include a commit hash, which will be used for testing. If you don't provide a commit hash, we will work off the default branch of your repository. Thus, if you plan on continuing work after delivery, we suggest you create a separate branch for either the delivery or your continuing work.
+>
+> If there is anything particular about any of the deliverables we or a future reader should know, use the respective `Notes` column.
+
+| Number | Deliverable | Link | Notes |
+| ------------- | ------------- | ------------- |------------- |
+| 0b. | Documentation |[architecture](https://github.com/nutsfinance/stable-asset/blob/affe16c004003307bfa9c5e3db733d78e1c9f33e/architecture.png)||
+| 0c.  | Testing |[xcm](https://github.com/nutsfinance/stable-asset/blob/affe16c004003307bfa9c5e3db733d78e1c9f33e/lib/stable-asset-xcm/src/tests.rs) [stable-asset](https://github.com/nutsfinance/stable-asset/blob/affe16c004003307bfa9c5e3db733d78e1c9f33e/lib/stable-asset/src/tests.rs)||
```
Comment (alxs) [inline deliveries/tdot-milestone-2.md:19]: Here to "comprehensive tests that covers minting tDOT locally in host chain and remotely in guest chain"

File: deliveries/tdot-milestone-3.md [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+* **Milestone Number:** 3
+
+> Please provide a list of all deliverables of the milestone extracted from the initial application and a link to the deliverable itself. Ideally all links inside the below table should include a commit hash, which will be used for testing. If you don't provide a commit hash, we will work off the default branch of your repository. Thus, if you plan on continuing work after delivery, we suggest you create a separate branch for either the delivery or your continuing work.
+>
+> If there is anything particular about any of the deliverables we or a future reader should know, use the respective `Notes` column.
+
+| Number | Deliverable | Link | Notes |
+| ------------- | ------------- | ------------- |------------- |
+| 0b. | Documentation |[architecture](https://nutsfinance.gitbook.io/tapio/)||
+| 0c.  | Testing |[xcm](https://github.com/nutsfinance/stable-asset/blob/affe16c004003307bfa9c5e3db733d78e1c9f33e/lib/stable-asset-xcm/src/tests.rs) [stable-asset](https://github.com/nutsfinance/stable-asset/blob/affe16c004003307bfa9c5e3db733d78e1c9f33e/lib/stable-asset/src/tests.rs)||
```
Comment (alxs) [inline deliveries/tdot-milestone-3.md:19]: And here "comprehensive tests that covers redeeming tDOT to host chain or to guest chains"

**Review by alxs — CHANGES_REQUESTED** [review]

File: deliveries/tdot-milestone-1.md (hunk 5 above)
Comment (ukby1234) [reply deliveries/tdot-milestone-1.md]: Updated

File: deliveries/tdot-milestone-2.md (hunk 6 above)
Comment (ukby1234) [reply deliveries/tdot-milestone-2.md]: Updated

File: deliveries/tdot-milestone-3.md (hunk 7 above)
Comment (ukby1234) [reply deliveries/tdot-milestone-3.md]: Updated

**Comment by alxs** [comment]
Many thanks @ukby1234. I still would like to ask you to please expand on these - the handful of tests you provide for each of these, e.g. 3 for milestone 1, can hardly be considered comprehensive.

**Comment by ukby1234** [comment]
> Many thanks @ukby1234. I still would like to ask you to please expand on these - the handful of tests you provide for each of these, e.g. 3 for milestone 1, can hardly be considered comprehensive.

Not sure what else can be added to tests though give we are hitting >=85% coverage unless you want to make that number even higher.

**Comment by ukby1234** [comment]
@alxs anyway, updated code to include more tests, which pushes the coverage to 89%.

**Comment by alxs** [comment]
Code coverage is only a metric and represents merely the quantity, but not the quality of the tests. High code coverage doesn't automatically mean that a piece of software is properly tested.

Thanks for the additional tests - these look great already. It would also be useful to have tests for the functionality you describe in the specification, e.g.:

>  If minting fails in host chain, the whole extrinsic is reverted. If minting fails in guest chain, user will get local LP.

**Comment by alxs** [comment]
By the way, I'm curious about the current tDOT [architecture](https://nutsfinance.gitbook.io/tapio/tapio/tdot#how-tdot-works): if I remember correctly, ACoconut and Stable Asset were based on a basket of equally valued assets. However, it looks like tDOT still maintains separate liquidity pools for each asset (which misses out on what I thought was the main benefit of Stable Swap, the aggregated liquidity), and furthermore these assets are not equally valued. As far as I'm aware, the two assets you use as examples both aren't and their value in relation to the peg fluctuates over time. How does this work at all? Is it still the case that the amount of tDOT in circulation is equal to the sum of all assets across all pools, and what is its value then? How does one redeem for assets of different value?

**Comment by ukby1234** [comment]
So let me explain this in details. Stable Asset pallet here are responsible for the stable swap liquidity pools, which can exist on any parachain, and the pool token is pegged to DOT in this case. Then we have another pallet, the `xcm` pallet, that mints 1:1 with any stable asset pallet pool tokens, subject to the minting limit, which is also pegged to DOT. Let's say users want to swap LcDOT on parachain A with DOT on parachain B. They can first mint the tDOT with LcDOT pool on parachain A, then redeem the tDOT tokens on parachain B with DOT. 
> Is it still the case that the amount of tDOT in circulation is equal to the sum of all assets across all pools, and what is its value then? 
Not always, when minting tDOT fails, users are expected to receive the stable asset pool token. So this means tDOT in circulation is less than or equal to the sum of all stable asset pool tokens. 
@shengda can add more details to this as well.

**Comment by ukby1234** [comment]
@alxs With respect to this test (If minting fails in host chain, the whole extrinsic is reverted. If minting fails in guest chain, user will get local LP.), I think tests should only be testing `lib.rs` not mocks in `mock.rs`. So with the scenario you described, these are actually asynchronous XCM calls. For example, when users want to mint tDOT, it calls [this function](https://github.com/nutsfinance/stable-asset/blob/e5c8c1ba19c730257e01077ce1c326476c1002c2/lib/stable-asset/src/lib.rs#L767), which sends the XCM message to call [this function](https://github.com/nutsfinance/stable-asset/blob/e5c8c1ba19c730257e01077ce1c326476c1002c2/lib/stable-asset-xcm/src/lib.rs#L442). Then when `mint` in `stable-asset-xcm` fails, it triggers [this function](https://github.com/nutsfinance/stable-asset/blob/e5c8c1ba19c730257e01077ce1c326476c1002c2/lib/stable-asset/src/lib.rs#L780) through XCM. 
Those XCM handling are traits like [this](https://github.com/nutsfinance/stable-asset/blob/e5c8c1ba19c730257e01077ce1c326476c1002c2/lib/stable-asset/src/lib.rs#L319) and [this](https://github.com/nutsfinance/stable-asset/blob/e5c8c1ba19c730257e01077ce1c326476c1002c2/lib/stable-asset-xcm/src/lib.rs#L125). This means testing those traits are not something can be done via unit tests. 
Also, the [implementations](https://github.com/nutsfinance/stable-asset/blob/e5c8c1ba19c730257e01077ce1c326476c1002c2/demo/runtime/src/lib.rs#L431-L520) in the demo chain doesn't utilize XCM at all so it's not production code and shouldn't write tests as well. The actual production-ready implementations will be like [this](https://github.com/AcalaNetwork/Acala/blob/ad240e9b96d4338a66fe7daad5bf53d8bb6a25f8/runtime/karura/src/lib.rs#L1627-L1774) and [this](https://github.com/nutsfinance/bifrost/blob/f0cba77760cf7e9b4576f6a255c6496edd36aad0/runtime/bifrost-kusama/src/lib.rs#L1976-L1997). 

The JavaScript client demonstrates this [capabilities](https://github.com/nutsfinance/stable-asset/blob/e5c8c1ba19c730257e01077ce1c326476c1002c2/demo/client/index.js#L89-L102) without using XCM just to show how those pallets are supposed to interact. 

Let me know if you have more questions.

**Comment by shengda** [comment]
> By the way, I'm curious about the current tDOT [architecture](https://nutsfinance.gitbook.io/tapio/tapio/tdot#how-tdot-works): if I remember correctly, ACoconut and Stable Asset were based on a basket of equally valued assets. However, it looks like tDOT still maintains separate liquidity pools for each asset (which misses out on what I thought was the main benefit of Stable Swap, the aggregated liquidity), and furthermore these assets are not equally valued. As far as I'm aware, the two assets you use as examples both aren't and their value in relation to the peg fluctuates over time. How does this work at all? Is it still the case that the amount of tDOT in circulation is equal to the sum of all assets across all pools, and what is its value then? How does one redeem for assets of different value?

The new architecture mains separate pools for different derivates, and these derivates can be swapped with the aggregated LP as medium so the liquidity is not segregated.

Each stable swap pool has its own parameters such as amplification parameters. These parameters controls the "shape" of the pool, i.e. the range of value change of each underlying derivative, but the overall value of each pool remains constant since its collateral factor will always increase if the underlying value of derivative change.

**Comment by ukby1234** [comment]
@alxs friendly bump

**Comment by alxs** [comment]
Hey @ukby1234 @shengda, sorry for the long delay. I had some backlog but I'll make sure to get back to you in time in the future.

Thanks for helping me understand this. I'm honestly still confused. As I mentioned, lcDOT and LDOT aren't actually pegged to DOT as far as I'm aware. They will always have a _similar_ value, but they are not 1:1 pegged and their value will fluctuate over time.

> Then we have another pallet, the xcm pallet, that mints 1:1 with any stable asset pallet pool tokens, subject to the minting limit, which is also pegged to DOT. Let's say users want to swap LcDOT on parachain A with DOT on parachain B. They can first mint the tDOT with LcDOT pool on parachain A, then redeem the tDOT tokens on parachain B with DOT.

In this example, the user should get a smaller amount of DOT than the lcDOT they had. But if, as you say, the minting is 1:1, this doesn't make sense.

> the overall value of each pool remains constant since its collateral factor will always increase if the underlying value of derivative change.

I've always been curious about this assertion you make. How can this be the case if the value of one of the assets in the pool increases?

@ukby1234 thanks for the explanation regarding the tests. That is fine by me, but wouldn't it be possible to write integration tests for the XCM calls?

**Comment by ukby1234** [comment]
@alxs It will be extremely hard to write tests to test the actual XCM invocations because of its asynchronous nature. However, I added some unit tests to cover those XCM stubs are called with desired parameters. 

I will defer the rest of the top-level questions to @shengda.

**Comment by ukby1234** [comment]
@alxs Let me try to answer those questions because @shengda is out for a few days. So the original `stable-asset` pallet is supposed to generate a token that is pegged to DOT. The way we handle this is through getting the exchange rate, let's say for LDOT through the actual staking process, convert them to the equivalent amount of DOT, and do calculation based on the current exchange rate. The relevant code is [here](https://github.com/AcalaNetwork/Acala/blob/master/runtime/karura/src/lib.rs#L1597-L1626) and the rebase token code is [here](https://github.com/open-web3-stack/open-runtime-module-library/blob/20969f385bf77e41386e8d8aa9626b003a306204/tokens/src/impls.rs#L8). 

The XCM pallet is based on those LP tokens of the original `stable-asset` pallet, which is designed to peg DOT in this case. Then when users want to exchange the token generated from the XCM pallet, they can simply redeem the XCM token to the local `stable-asset` token, and then redeem the local LP token to their desired xDOT tokens. 

Since calculations on how many xDOT received for redemption of XCM pallet token are done using the `stable-asset` pallet, users should receive a correct amount of xDOT.

**Comment by ukby1234** [comment]
@alxs friendly bump

**Comment by alxs** [comment]
@ukby1234 thanks for the ping and really thanks a lot for the detailed explanation once again. I get what you're saying and now the code and your earlier comments make more sense, but I get the impression that then the system really depends on this one exchange rate sourcing method and not on the fact that the assets are meant to be equally valued, as per Stable Asset's original design.

I wonder how you will determine the price for other assets for which there is no such mechanism and whose price is only determined by the exchange rate (I think that would be the case for lcDOT?). I hope you're careful and thoroughly audit this and future additions before they go live.

In any case, I can confirm that your delivery meets the specifications in your application and is hence accepted. You can find my delivery notes here: [1](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/tdot_1_alxs.md), [2](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/tdot_2_alxs.md), [3](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/tdot_3_alxs.md). I'll notify the finance team to pay out your invoice - as usual, please allow for up to 2 weeks of processing time.

Thanks a lot again for your efforts and best of luck with the continuation of the project.

**Pull request merged**